File: app/src/main/java/jp/hazuki/yuzubrowser/utils/view/pie/PieMenu.java
/*
 * Copyright (C) 2017 Hazuki
 * fixed by Cynthia Project
 * Copyright (C) 2010 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *	  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.utils.view.pie;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Point;
import android.graphics.PointF;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.graphics.RectF;
import android.support.v4.content.res.ResourcesCompat;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.SoundEffectConstants;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import java.util.ArrayList;
import java.util.List;

import jp.hazuki.yuzubrowser.R;

public class PieMenu extends FrameLayout {

    private static final int MAX_LEVELS = 5;

    public interface PieController {
        /**
         * called before menu opens to customize menu
         * returns if pie state has been changed
         */
        boolean onOpen();
    }

    /**
     * A view like object that lives off of the pie menu
     */
    public interface PieView {

        interface OnLayoutListener {
            void onLayout(int ax, int ay, boolean left);
        }

        void setLayoutListener(OnLayoutListener l);

        void layout(int anchorX, int anchorY, boolean onleft, float angle);

        void draw(Canvas c);

        boolean onTouchEvent(MotionEvent evt);

    }

    private Point mCenter;
    private int mRadius;
    private int mRadiusInc;
    private int mSlop;
    private int mTouchOffset;

    private int mPosition;
    private boolean mOpen;
    private PieController mController;

    private List<PieItem> mItems;
    private int mLevels;
    private int[] mCounts;
    private PieView mPieView = null;

    private Paint mNormalPaint;
    private Paint mSelectedPaint;

    // touch handling
    PieItem mCurrentItem;

    /**
     * @param context
     * @param attrs
     * @param defStyle
     */
    public PieMenu(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(context);
    }

    /**
     * @param context
     * @param attrs
     */
    public PieMenu(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    /**
     * @param context
     */
    public PieMenu(Context context) {
        super(context);
        init(context);
    }

    private void init(Context ctx) {
        mItems = new ArrayList<>();
        mLevels = 0;
        mCounts = new int[MAX_LEVELS];
        Resources res = ctx.getResources();
        mRadius = (int) res.getDimension(R.dimen.qc_radius_start);
        mRadiusInc = (int) res.getDimension(R.dimen.qc_radius_increment);
        mSlop = (int) res.getDimension(R.dimen.qc_slop);
        mTouchOffset = (int) res.getDimension(R.dimen.qc_touch_offset);
        mOpen = false;
        setWillNotDraw(false);
        setDrawingCacheEnabled(false);
        mCenter = new Point(0, 0);
        mNormalPaint = new Paint();
        mNormalPaint.setColor(ResourcesCompat.getColor(res, R.color.qc_normal, ctx.getTheme()));
        mNormalPaint.setAntiAlias(true);
        mSelectedPaint = new Paint();
        mSelectedPaint.setColor(ResourcesCompat.getColor(res, R.color.qc_selected, ctx.getTheme()));
        mSelectedPaint.setAntiAlias(true);
    }

    public void setController(PieController ctl) {
        mController = ctl;
    }

    public void addItem(PieItem item) {
        // add the item to the pie itself
        mItems.add(item);
        int l = item.getLevel();
        mLevels = Math.max(mLevels, l);
        mCounts[l]++;
    }

    public void removeItem(PieItem item) {
        mItems.remove(item);
    }

    public void clearItems() {
        mItems.clear();
    }

    private boolean onTheLeft() {
        return mCenter.x < mSlop;
    }

    /**
     * guaranteed has center set
     *
     * @param show
     */
    private void show(boolean show) {
        mOpen = show;
        if (mOpen) {
            if (mController != null) {
                mController.onOpen();
            }
            layoutPie();
        }
        if (!show) {
            mCurrentItem = null;
            mPieView = null;
        }
        invalidate();
    }

    private void setCenter(int x, int y) {
        if (x < mSlop) {

            mCenter.x = 0;
        } else {
            mCenter.x = getWidth();
        }
        mCenter.y = y;
    }

    private void layoutPie() {
        float emptyangle = (float) Math.PI / 16;
        int rgap = 2;
        int inner = mRadius + rgap;
        int outer = mRadius + mRadiusInc - rgap;
        int gap = 1;
        for (int i = 0; i < mLevels; i++) {
            int level = i + 1;
            float sweep = (float) (Math.PI - 2 * emptyangle) / mCounts[level];
            float angle = emptyangle + sweep / 2;
            for (PieItem item : mItems) {
                if (item.getLevel() == level) {
                    View view = item.getView();
                    int w = view.getLayoutParams().width;
                    int h = view.getLayoutParams().height;
                    int r = inner + (outer - inner) * 55 / 100;
                    int x = (int) (r * Math.sin(angle));
                    int y = mCenter.y - (int) (r * Math.cos(angle)) - h / 2;
                    if (onTheLeft()) {
                        x = mCenter.x + x - w / 2;
                    } else {
                        x = mCenter.x - x - w / 2;
                    }
                    view.layout(x, y, x + w, y + h);
                    float itemstart = angle - sweep / 2;
                    Path slice = makeSlice(getDegrees(itemstart) - gap,
                            getDegrees(itemstart + sweep) + gap,
                            outer, inner, mCenter);
                    item.setGeometry(itemstart, sweep, inner, outer, slice);
                    angle += sweep;
                }
            }
            inner += mRadiusInc;
            outer += mRadiusInc;
        }
    }


    /**
     * converts a
     *
     * @param angle from 0..PI to Android degrees (clockwise starting at 3
     *              o'clock)
     * @return skia angle
     */
    private float getDegrees(double angle) {
        return (float) (270 - 180 * angle / Math.PI);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        if (mOpen) {
            int state;
            for (PieItem item : mItems) {
                Paint p = item.isSelected() ? mSelectedPaint : mNormalPaint;
                state = canvas.save();
                if (onTheLeft()) {
                    canvas.scale(-1, 1);
                }
                drawPath(canvas, item.getPath(), p);
                canvas.restoreToCount(state);
                drawItem(canvas, item);
            }
            if (mPieView != null) {
                mPieView.draw(canvas);
            }
        }
    }

    private void drawItem(Canvas canvas, PieItem item) {
        // draw the item view
        View view = item.getView();
        int state = canvas.save();
        canvas.translate(view.getX(), view.getY());
        view.draw(canvas);
        canvas.restoreToCount(state);
    }

    private void drawPath(Canvas canvas, Path path, Paint paint) {
        canvas.drawPath(path, paint);
    }

    private Path makeSlice(float start, float end, int outer, int inner, Point center) {
        RectF bb =
                new RectF(center.x - outer, center.y - outer, center.x + outer,
                        center.y + outer);
        RectF bbi =
                new RectF(center.x - inner, center.y - inner, center.x + inner,
                        center.y + inner);
        Path path = new Path();
        path.arcTo(bb, start, end - start, true);
        path.arcTo(bbi, end, start - end);
        path.close();
        return path;
    }

    // touch handling for pie

    @Override
    public boolean onTouchEvent(MotionEvent evt) {
        float x = evt.getX();
        float y = evt.getY();
        int action = evt.getActionMasked();
        if (MotionEvent.ACTION_DOWN == action) {
            if ((mPosition != 1 && x > getWidth() - mSlop) || (mPosition != 2 && x < mSlop)) {
                setCenter((int) x, (int) y);
                show(true);
                return true;
            }
        } else if (MotionEvent.ACTION_UP == action) {
            if (mOpen) {
                boolean handled = false;
                if (mPieView != null) {
                    handled = mPieView.onTouchEvent(evt);
                }
                PieItem item = mCurrentItem;
                deselect();
                show(false);
                if (!handled && (item != null)) {
                    item.getView().performClick();
                }
                return true;
            }
        } else if (MotionEvent.ACTION_CANCEL == action) {
            if (mOpen) {
                show(false);
            }
            deselect();
            return false;
        } else if (MotionEvent.ACTION_MOVE == action) {
            boolean handled = false;
            PointF polar = getPolar(x, y);
            int maxr = mRadius + mLevels * mRadiusInc + 50;
            if (mPieView != null) {
                handled = mPieView.onTouchEvent(evt);
            }
            if (handled) {
                invalidate();
                return false;
            }
            if (polar.y > maxr) {
                deselect();
                show(false);
                evt.setAction(MotionEvent.ACTION_DOWN);
                if (getParent() != null) {
                    ((ViewGroup) getParent()).dispatchTouchEvent(evt);
                }
                return false;
            }
            PieItem item = findItem(polar);
            if (mCurrentItem != item) {
                onEnter(item);
                if ((item != null) && item.isPieView()) {
                    int cx = item.getView().getLeft() + (onTheLeft()
                            ? item.getView().getWidth() : 0);
                    int cy = item.getView().getTop();
                    mPieView = item.getPieView();
                    layoutPieView(mPieView, cx, cy,
                            (item.getStartAngle() + item.getSweep()) / 2);
                }
                invalidate();
            }
        }
        // always re-dispatch event
        return false;
    }

    private void layoutPieView(PieView pv, int x, int y, float angle) {
        pv.layout(x, y, onTheLeft(), angle);
    }

    /**
     * enter a slice for a view
     * updates model only
     *
     * @param item
     */
    private void onEnter(PieItem item) {
        // deselect
        if (mCurrentItem != null) {
            mCurrentItem.setSelected(false);
        }
        if (item != null) {
            // clear up stack
            playSoundEffect(SoundEffectConstants.CLICK);
            item.setSelected(true);
            mPieView = null;
        }
        mCurrentItem = item;
    }

    private void deselect() {
        if (mCurrentItem != null) {
            mCurrentItem.setSelected(false);
        }
        mCurrentItem = null;
        mPieView = null;
    }

    private PointF getPolar(float x, float y) {
        PointF res = new PointF();
        // get angle and radius from x/y
        res.x = (float) Math.PI / 2;
        x = mCenter.x - x;
        if (mCenter.x < mSlop) {
            x = -x;
        }
        y = mCenter.y - y;
        res.y = (float) Math.sqrt(x * x + y * y);
        if (y > 0) {
            res.x = (float) Math.asin(x / res.y);
        } else if (y < 0) {
            res.x = (float) (Math.PI - Math.asin(x / res.y));
        }
        return res;
    }

    /**
     * @param polar x: angle, y: dist
     * @return the item at angle/dist or null
     */
    private PieItem findItem(PointF polar) {
        // find the matching item:
        for (PieItem item : mItems) {
            if ((item.getInnerRadius() - mTouchOffset < polar.y)
                    && (item.getOuterRadius() - mTouchOffset > polar.y)
                    && (item.getStartAngle() < polar.x)
                    && (item.getStartAngle() + item.getSweep() > polar.x)) {
                return item;
            }
        }
        return null;
    }

    public void notifyChangeState() {
        for (PieItem item : mItems) {
            item.notifyChangeState();
        }
    }

    public void setRadiusStart(int radius) {
        this.mRadius = radius;
    }

    public void setRadiusIncrement(int radiusInc) {
        this.mRadiusInc = radiusInc;
    }

    public void setSlop(int slop) {
        this.mSlop = slop;
    }

    public void setPosition(int position) {
        this.mPosition = position;
    }

    public void setTouchOffset(int touchOffset) {
        this.mTouchOffset = touchOffset;
    }

    public void setNormalColor(int color) {
        mNormalPaint.setColor(color);
    }

    public void setSelectedColor(int color) {
        mSelectedPaint.setColor(color);
    }

    public void setColorFilterToItems(int color) {
        ColorFilter cf = (color != 0) ? new PorterDuffColorFilter(color, PorterDuff.Mode.MULTIPLY) : null;
        for (PieItem item : mItems) {
            item.getView().setColorFilter(cf);
        }
    }
}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/utils/view/pie/PieMenu.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.utils.view.pie

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.SoundEffectConstants
import android.view.ViewGroup
import android.widget.FrameLayout
import jp.hazuki.yuzubrowser.core.utility.extensions.dimension
import jp.hazuki.yuzubrowser.core.utility.extensions.getResColor
import jp.hazuki.yuzubrowser.legacy.R
import java.util.*

class PieMenu @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr) {

    private val center = Point(0, 0)
    private var radius: Int = 0
    private var radiusInc: Int = 0
    private var slop: Int = 0
    private var touchOffset: Int = 0

    private var position: Int = 0
    var isOpen: Boolean = false
        private set
    private var controller: PieController? = null

    private val items = ArrayList<PieItem>()
    private var levels: Int = 0
    private val counts = IntArray(MAX_LEVELS)
    private var pieView: PieView? = null

    private val normalPaint = Paint()
    private val selectedPaint = Paint()

    // touch handling
    internal var currentItem: PieItem? = null

    interface PieController {
        /**
         * called before menu opens to customize menu
         * returns if pie state has been changed
         */
        fun onOpen(): Boolean
    }

    /**
     * A view like object that lives off of the pie menu
     */
    interface PieView {

        interface OnLayoutListener {
            fun onLayout(ax: Int, ay: Int, left: Boolean)
        }

        fun setLayoutListener(l: OnLayoutListener)

        fun layout(anchorX: Int, anchorY: Int, onleft: Boolean, angle: Float)

        fun draw(c: Canvas)

        fun onTouchEvent(evt: MotionEvent): Boolean

    }

    init {
        radius = context.dimension(R.dimen.qc_radius_start)
        radiusInc = context.dimension(R.dimen.qc_radius_increment)
        slop = context.dimension(R.dimen.qc_slop)
        touchOffset = context.dimension(R.dimen.qc_touch_offset)
        setWillNotDraw(false)
        isDrawingCacheEnabled = false
        normalPaint.color = context.getResColor(R.color.qc_normal)
        normalPaint.isAntiAlias = true
        selectedPaint.color = context.getResColor(R.color.qc_selected)
        selectedPaint.isAntiAlias = true
    }

    fun setController(ctl: PieController) {
        controller = ctl
    }

    fun addItem(item: PieItem) {
        // add the item to the pie itself
        items.add(item)
        val l = item.level
        levels = Math.max(levels, l)
        counts[l]++
    }

    fun removeItem(item: PieItem) {
        items.remove(item)
        counts[item.level]--
    }

    fun clearItems() {
        items.clear()
        levels = 0
        for (i in 0 until MAX_LEVELS) {
            counts[i] = 0
        }
    }

    private fun onTheLeft(): Boolean {
        return center.x < slop
    }

    /**
     * guaranteed has center set
     *
     * @param show
     */
    private fun show(show: Boolean) {
        isOpen = show
        if (isOpen) {
            if (controller != null) {
                controller!!.onOpen()
            }
            layoutPie()
        }
        if (!show) {
            currentItem = null
            pieView = null
        }
        invalidate()
    }

    private fun setCenter(x: Int, y: Int) {
        if (x < slop) {

            center.x = 0
        } else {
            center.x = width
        }
        center.y = y
    }

    private fun layoutPie() {
        val emptyangle = Math.PI.toFloat() / 16
        val rgap = 2
        var inner = radius + rgap
        var outer = radius + radiusInc - rgap
        val gap = 1
        for (i in 0 until levels) {
            val level = i + 1
            val sweep = (Math.PI - 2 * emptyangle).toFloat() / counts[level]
            var angle = emptyangle + sweep / 2
            for (item in items) {
                if (item.level == level) {
                    val view = item.view
                    val w = view.layoutParams.width
                    val h = view.layoutParams.height
                    val r = inner + (outer - inner) * 55 / 100
                    var x = (r * Math.sin(angle.toDouble())).toInt()
                    val y = center.y - (r * Math.cos(angle.toDouble())).toInt() - h / 2
                    x = if (onTheLeft()) {
                        center.x + x - w / 2
                    } else {
                        center.x - x - w / 2
                    }
                    view.layout(x, y, x + w, y + h)
                    val itemstart = angle - sweep / 2
                    val slice = makeSlice(getDegrees(itemstart.toDouble()) - gap,
                            getDegrees((itemstart + sweep).toDouble()) + gap,
                            outer, inner, center)
                    item.setGeometry(itemstart, sweep, inner, outer, slice)
                    angle += sweep
                }
            }
            inner += radiusInc
            outer += radiusInc
        }
    }


    /**
     * converts a
     *
     * @param angle from 0..PI to Android degrees (clockwise starting at 3
     * o'clock)
     * @return skia angle
     */
    private fun getDegrees(angle: Double): Float {
        return (270 - 180 * angle / Math.PI).toFloat()
    }

    override fun onDraw(canvas: Canvas) {
        if (isOpen) {
            var state: Int
            for (item in items) {
                val p = if (item.isSelected) selectedPaint else normalPaint
                state = canvas.save()
                if (onTheLeft()) {
                    canvas.scale(-1f, 1f)
                }
                drawPath(canvas, item.path, p)
                canvas.restoreToCount(state)
                drawItem(canvas, item)
            }
            if (pieView != null) {
                pieView!!.draw(canvas)
            }
        }
    }

    private fun drawItem(canvas: Canvas, item: PieItem) {
        // draw the item view
        val view = item.view
        val state = canvas.save()
        canvas.translate(view.x, view.y)
        view.draw(canvas)
        canvas.restoreToCount(state)
    }

    private fun drawPath(canvas: Canvas, path: Path?, paint: Paint?) {
        canvas.drawPath(path!!, paint!!)
    }

    private fun makeSlice(start: Float, end: Float, outer: Int, inner: Int, center: Point?): Path {
        val bb = RectF((center!!.x - outer).toFloat(), (center.y - outer).toFloat(), (center.x + outer).toFloat(),
                (center.y + outer).toFloat())
        val bbi = RectF((center.x - inner).toFloat(), (center.y - inner).toFloat(), (center.x + inner).toFloat(),
                (center.y + inner).toFloat())
        val path = Path()
        path.arcTo(bb, start, end - start, true)
        path.arcTo(bbi, end, start - end)
        path.close()
        return path
    }

    // touch handling for pie

    override fun onTouchEvent(evt: MotionEvent): Boolean {
        val x = evt.x
        val y = evt.y
        val action = evt.actionMasked
        if (MotionEvent.ACTION_DOWN == action) {
            if (position != 1 && x > width - slop || position != 2 && x < slop) {
                setCenter(x.toInt(), y.toInt())
                show(true)
                return true
            }
        } else if (MotionEvent.ACTION_UP == action) {
            if (isOpen) {
                val handled = pieView?.onTouchEvent(evt) ?: false
                val item = currentItem
                deselect()
                show(false)
                if (!handled && item != null) {
                    item.view.performClick()
                }
                return true
            }
        } else if (MotionEvent.ACTION_CANCEL == action) {
            if (isOpen) {
                show(false)
            }
            deselect()
            return false
        } else if (MotionEvent.ACTION_MOVE == action) {
            val handled = pieView?.onTouchEvent(evt) ?: false
            val polar = getPolar(x, y)
            val maxr = radius + levels * radiusInc + 50
            if (handled) {
                invalidate()
                return false
            }
            if (polar.y > maxr) {
                deselect()
                show(false)
                evt.action = MotionEvent.ACTION_DOWN
                if (parent != null) {
                    (parent as ViewGroup).dispatchTouchEvent(evt)
                }
                return false
            }
            val item = findItem(polar)
            if (currentItem != item) {
                onEnter(item)
                if (item != null && item.isPieView) {
                    val cx = item.view.left + if (onTheLeft())
                        item.view.width
                    else
                        0
                    val cy = item.view.top
                    pieView = item.pieView
                    layoutPieView(pieView!!, cx, cy,
                            (item.startAngle + item.sweep) / 2)
                }
                invalidate()
            }
        }
        // always re-dispatch event
        return false
    }

    private fun layoutPieView(pv: PieView, x: Int, y: Int, angle: Float) {
        pv.layout(x, y, onTheLeft(), angle)
    }

    /**
     * enter a slice for a view
     * updates model only
     *
     * @param item
     */
    private fun onEnter(item: PieItem?) {
        // deselect
        currentItem?.isSelected = false
        if (item != null) {
            // clear up stack
            playSoundEffect(SoundEffectConstants.CLICK)
            item.isSelected = true
            pieView = null
        }
        currentItem = item
    }

    private fun deselect() {
        currentItem?.isSelected = false
        currentItem = null
        pieView = null
    }

    private fun getPolar(tx: Float, ty: Float): PointF {
        var x = tx
        var y = ty
        val res = PointF()
        // get angle and radius from x/y
        res.x = Math.PI.toFloat() / 2
        x = center.x - x
        if (center.x < slop) {
            x = -x
        }
        y = center.y - y
        res.y = Math.sqrt((x * x + y * y).toDouble()).toFloat()
        if (y > 0) {
            res.x = Math.asin((x / res.y).toDouble()).toFloat()
        } else if (y < 0) {
            res.x = (Math.PI - Math.asin((x / res.y).toDouble())).toFloat()
        }
        return res
    }

    /**
     * @param polar x: angle, y: dist
     * @return the item at angle/dist or null
     */
    private fun findItem(polar: PointF): PieItem? {
        // find the matching item:
        return items.firstOrNull {
            it.innerRadius - touchOffset < polar.y
                    && it.outerRadius - touchOffset > polar.y
                    && it.startAngle < polar.x
                    && it.startAngle + it.sweep > polar.x
        }
    }

    fun notifyChangeState() {
        for (item in items) {
            item.notifyChangeState()
        }
    }

    fun setRadiusStart(radius: Int) {
        this.radius = radius
    }

    fun setRadiusIncrement(radiusInc: Int) {
        this.radiusInc = radiusInc
    }

    fun setSlop(slop: Int) {
        this.slop = slop
    }

    fun setPosition(position: Int) {
        this.position = position
    }

    fun setTouchOffset(touchOffset: Int) {
        this.touchOffset = touchOffset
    }

    fun setNormalColor(color: Int) {
        normalPaint.color = color
    }

    fun setSelectedColor(color: Int) {
        selectedPaint.color = color
    }

    fun setColorFilterToItems(color: Int) {
        val cf = if (color != 0) PorterDuffColorFilter(color, PorterDuff.Mode.MULTIPLY) else null
        for (item in items) {
            item.view.colorFilter = cf
        }
    }

    companion object {

        private const val MAX_LEVELS = 5
    }
}
